packages/cubejs-client-react/src/CubeProvider.jsx

```diff
@@ -1,6 +1,16 @@
-import React from 'react';
+import React, { useContext, useMemo } from 'react';
 import CubeContext from './CubeContext';
 
-export default function CubeProvider({ cubejsApi, children }) {
-  return <CubeContext.Provider value={{ cubejsApi }}>{children}</CubeContext.Provider>;
+export default function CubeProvider({
+  name,
+  cubejsApi,
+  children,
+}) {
+  const oldContext = useContext(CubeContext);
+  const newContext = useMemo(() => ({
+    ...oldContext,
+    [name ?? 'default']: cubejsApi,
+  }), []);
```

Comment:
It won't work without `[oldContext]`

Also, I'm not sure about this approach as it seems a bit unnatural having multiple `CubeProvider`s override the context value.

I would rather make the `value` an object, e.g.

```jsx
<CubeProvider = {{
  myApi: cubeApi1,
  myOtherApi: cubeApi2
  // ...
```

Reply:
I'm not sure I understand the example.

As currently implemented, you would write:

```tsx
<CubeProvider name="foo" cubeJsApi={fooApi}>
  <CubeProvider name="bar" cubeJsApi={barApi}>
    <CubeProvider cubeJsApi={defaultApi}> {/* name is "default" */}
      {…}
    </CubeProvider>
  </CubeProvider>
</CubeProvider>
```

Internally, the APIs are kept in a shared context, but they wouldn't override each other as long as they have different names. The fact that all of the APIs are stored in a single object in the context is an implementation detail.

We could do the same thing with a different API:

```tsx
<CubeProvider cubeJsApi={{
  foo: fooApi,
  bar: barApi,
  default: defaultApi,
}}>
  {…}
</CubeProvider>
```

I like the `name` prop myself, but this API would be fine. The behavior should be identical and any performance differences negligible.

Also, `...oldContext` should work fine; it's object spread, not array spread, so it will work even if `oldContext` is null. I might be misunderstanding what you're saying here.

Reply:
I've updated the example where I left out the `cubejsApi=` part.

While the approach with `name` works just fine, it's more error-prone (e.g. duplicate, missing names). Also, it's feels a bit off because technically you can access a `cubeApi` outside of its own provider (because we have multiple providers and a single context)

> Also, ...oldContext should work fine; it's object spread, not array spread, so it will work even if oldContext is null. I might be misunderstanding what you're saying here.

I meant you needed to include it as the useEffect's dependency otherwise, it won't update